pytensor/tensor/math.py

```diff
@@ -1571,6 +1574,63 @@ def std(input, axis=None, ddof=0, keepdims=False, corrected=False):
     return ret
 
 
+def median(input, axis=None):
+    """
+    Computes the median along the given axis(es) of a tensor `input`.
+
+    Parameters
+    ----------
+    input: TensorVariable
+        The input tensor.
+    axis: None or int or (list of int) (see `Sum`)
+        Compute the median along this axis of the tensor.
+        None means all axes (like numpy).
+    """
+    from pytensor.ifelse import ifelse
+
+    input = as_tensor_variable(input)
+    input_ndim = input.type.ndim
+    if axis is None:
+        axis = list(range(input_ndim))
+    elif isinstance(axis, int | np.integer):
+        axis = [axis]
+    elif isinstance(axis, np.ndarray) and axis.ndim == 0:
+        axis = [int(axis)]
+    else:
+        axis = [int(a) for a in axis]
+
+    new_axes_order = [i for i in range(input.ndim) if i not in axis] + list(axis)
+    input = input.dimshuffle(new_axes_order)
+    input_shape = input.shape
+
+    remaining_axis_size = shape(input)[: input.ndim - len(axis)]
+    flattened_axis_size = prod(shape(input)[input.ndim - len(axis) :])
+
+    input = input.reshape(concatenate([remaining_axis_size, [flattened_axis_size]]))
+    axis = -1
+
+    # Sort the input tensor along the specified axis
+    sorted_input = input.sort(axis=axis)
+    input_shape = input.shape[axis]
+    k = extract_constant(input_shape) // 2
+
+    indices1 = expand_dims(full_like(sorted_input.take(0, axis=axis), k - 1), axis)
+    indices2 = expand_dims(full_like(sorted_input.take(0, axis=axis), k), axis)
```

Comment:
What is happening here? Can't quite follow

Reply:
This was a tricky part. 
The problem started when ```indices``` was used in ```take_along_axis```. We need to ensure that the shape of the indices tensor (indices1) matches the shape required for broadcasting during the selection of elements using take_along_axis (shape of sorted_inputs). 
The ```full_like``` essentially makes an array of k/k-1 (central element indices) with the shape similar to the shape of the tensor along that axis( calculated using ```sorted_input.take(0, axis=axis)```.

I think this will be better with an example:
```
sorted_input = np.array([[1, 3, 5],
                         [2, 4, 6],
                         [7, 8, 9]])
k = sorted_input.shape[1] // 2  # Middle index for axis 1
# For this example, k = 3 // 2 = 1

first_elements = sorted_input.take(0, axis=1)
# first_elements = [1, 2, 7]

k_minus_1 = k - 1  # k - 1 = 0
full_tensor = full_like(first_elements, k_minus_1)
# full_tensor = [0, 0, 0]

indices1 = expand_dims(full_tensor, axis=1)
# indices1 = [[0],
#             [0],
#             [0]]

ans1 = take_along_axis(sorted_input, indices1, axis=axis)
# ans1 = [[1],
#         [2],
#         [7]]

k = k   # k = 1
full_tensor = full_like(first_elements, k)
# full_tensor = [1, 1, 1]

indices2 = expand_dims(full_tensor, axis=1)
# indices2 = [[1],
#             [1],
#             [1]]

ans2 = take_along_axis(sorted_input, indices2, axis=axis)
# ans2 = [[3],
#         [4],
#         [8]]
```
And hence based on ans1 and ans2, median is calculated.

Reply:
Okay so take_along_axis is not what we want, since we don't need advanced (array indexing). indices is a scalar and always the same. We can do `sorted_inputs[:, :, ..., :, k)` to get the value we want. Where the : are empty slices before the axis we want to index. I thought `take_along_axis` would do this, but apparently it's only for advanced (list of numbers) indexing

Reply:
For what basic vs advanced indexing means check out the numpy docs: https://numpy.org/doc/stable/user/basics.indexing.html